README.md
# pry-debugger-jruby

_JRuby 9k-compatible pry debugging!_

Using MRI? I strongly recommend [`pry-byebug`](https://github.com/deivid-rodriguez/pry-byebug) instead!

Adds `step`, `next`, `finish`, and `continue` commands and `breakpoints` to [Pry](http://pry.github.com).

To use, run JRuby with the `--debug` flag, and then invoke `pry` normally:

```ruby
def some_method
  binding.pry          # Execution will stop here.
  puts 'Hello, World!' # Run 'step' or 'next' in the console to move here.
end
```

You can also add the `--debug` flag to your `JRUBY_OPTS` environment variable, so it will be picked up by any ruby application. Do note that running `JRuby` in debug mode **does** have a noticeable impact on performance.

## Execution Commands

* `step`: Step execution into the next line or method. Takes an optional numeric argument to step multiple times.

* `next`: Step over to the next line within the same frame. Also takes an optional numeric argument to step multiple lines.

* `finish`: Execute until current stack frame returns.

* `continue`: Continue program execution and end the Pry session.

## Breakpoints

You can set and adjust breakpoints directly from a Pry session using the following commands:

* `break`: Set a new breakpoint from a line number in the current file, a file and line number, or a method. Pass an optional expression to create a conditional breakpoint. Edit existing breakpoints via various flags.

    Examples:

    ```ruby
break SomeClass#run            Break at the start of `SomeClass#run`.
break Foo#bar if baz?          Break at `Foo#bar` only if `baz?`.
break app/models/user.rb:15    Break at line 15 in user.rb.
break 14                       Break at line 14 in the current file.

break --condition 4 x > 2      Change condition on breakpoint #4 to 'x > 2'.
break --condition 3            Remove the condition on breakpoint #3.

break --delete 5               Delete breakpoint #5.
break --disable-all            Disable all breakpoints.

break                          List all breakpoints. (Same as `breakpoints`)
break --show 2                 Show details about breakpoint #2.
    ```

    Type `break --help` from a Pry session to see all available options.

* `breakpoints`: List all defined breakpoints. Pass `-v` or `--verbose` to see the source code around each breakpoint.

## Remote debugging

Support for [pry-remote](https://github.com/Mon-Ouie/pry-remote) is also included. Requires explicity requiring `pry-debugger-jruby`, not just relying on pry's plugin loader.

Want to debug a Rails app running inside `foreman`? Add to your `Gemfile`:

```ruby
gem 'pry'
gem 'pry-remote'
gem 'pry-debugger-jruby'
```

Then add `binding.remote_pry` where you want to pause:

```ruby
class UsersController < ApplicationController
  def index
    binding.remote_pry
    # ...
  end
end
```

Load a page that triggers the code. Connect to the session:

```bash
$ bundle exec pry-remote
```

## Tips

Stepping through code often? Add the following shortcuts to `~/.pryrc`:

```ruby
if defined?(PryDebuggerJRuby)
  Pry.commands.alias_command 'c', 'continue'
  Pry.commands.alias_command 's', 'step'
  Pry.commands.alias_command 'n', 'next'
  Pry.commands.alias_command 'f', 'finish'
end
```

## Contributors

`pry-debugger-jruby` is maintained by [@ivoanjo](https://github.com/ivoanjo/) and is based off the awesome previous work from the [`pry-debugger`](https://github.com/nixme/pry-debugger) creators:

* Gopal Patel (@nixme)
* John Mair (@banister)
* Nicolas Viennot (@nviennot)
* Benjamin R. Haskell (@benizi)
* Joshua Hou (@jshou)
* ...and others who helped with [`pry-nav`](https://github.com/nixme/pry-nav)

Patches and bug reports are welcome. Just send in a pull request or issue :)


CHANGELOG.md
## 1.1.0 (2016-12-02)

* Avoid stepping into pry code
* Add help text when running without `--debug`

## 1.0.0 (2016-12-01)

* Revamp documentation for JRuby usage
* Add workaround for using `step` in JRuby
* Fix circular require warning
* Fixed missing banner in commands causing uninitialized variable warning
* Drop support for ruby < 2.2
* Officially forked from `pry-debugger`, renaming gem to `pry-debugger-jruby` and releasing on RubyGems

## 0.2.3 (2014-07-02)

* Support [pry] 0.10

## 0.2.2 (2013-03-07)

* Relaxed [debugger][debugger] dependency.

## 0.2.1 (2012-12-26)

* Support breakpoints on methods defined in the pry console. (@banister)
* Fix support for specifying breakpoints by *file:line_number*. (@nviennot)
* Validate breakpoint conditionals are real Ruby expressions.
* Support for [debugger][debugger] ~> 1.2.0. (@jshou)
* Safer `alias_method_chain`-style patching of `Pry.start` and
  `PryRemote::Server#teardown`. (@benizi)

## 0.2.0 (2012-06-11)

* Breakpoints
* **finish** command
* Internal cleanup and bug fixes

## 0.1.0 (2012-06-07)

* First release. **step**, **next**, and **continue** commands.
  [pry-remote 0.1.4][pry-remote] support.


[pry]:         http://pryrepl.org/
[pry-remote]:  https://github.com/Mon-Ouie/pry-remote
[debugger]:    https://github.com/cldwalker/debugger


lib/pry-debugger-jruby/commands.rb
require 'pry' unless defined? Pry
require 'pry-debugger-jruby/breakpoints'

module PryDebuggerJRuby
  Commands = Pry::CommandSet.new do
    create_command 'step' do
      description 'Step execution into the next line or method.'

      banner <<-BANNER
        Usage: step [TIMES]

        Step execution forward. By default, moves a single step.

        Examples:

          step                           Move a single step forward.
          step 5                         Execute the next 5 steps.
      BANNER

      def process
        check_file_context
        breakout_navigation :step, args.first
      end
    end

    create_command 'next' do
      description 'Execute the next line within the current stack frame.'

      banner <<-BANNER
        Usage: next [LINES]

        Step over within the same frame. By default, moves forward a single
        line.

        Examples:

          next                           Move a single line forward.
          next 4                         Execute the next 4 lines.
      BANNER

      def process
        check_file_context
        breakout_navigation :next, args.first
      end
    end

    create_command 'finish' do
      description 'Execute until current stack frame returns.'

      banner <<-BANNER
        Usage: finish

        Execute until current stack frame returns.
      BANNER

      def process
        check_file_context
        breakout_navigation :finish
      end
    end

    create_command 'continue' do
      description 'End the Pry session and continue program execution.'

      banner <<-BANNER
        Usage: finish

        End the Pry session and continue program execution.
      BANNER

      def process
        check_file_context
        run 'exit-all'
      end
    end

    create_command 'break' do
      description 'Set or edit a breakpoint.'

      banner <<-BANNER
        Usage:   break <METHOD | FILE:LINE | LINE> [if CONDITION]
                 break --condition N [CONDITION]
                 break [--show | --delete | --enable | --disable] N
                 break [--delete-all | --disable-all]
        Aliases: breakpoint

        Set a breakpoint. Accepts a line number in the current file, a file and
        line number, or a method, and an optional condition.

        Pass appropriate flags to manipulate existing breakpoints.

        Examples:

          break SomeClass#run            Break at the start of `SomeClass#run`.
          break Foo#bar if baz?          Break at `Foo#bar` only if `baz?`.
          break app/models/user.rb:15    Break at line 15 in user.rb.
          break 14                       Break at line 14 in the current file.

          break --condition 4 x > 2      Add/change condition on breakpoint #4.
          break --condition 3            Remove the condition on breakpoint #3.

          break --delete 5               Delete breakpoint #5.
          break --disable-all            Disable all breakpoints.

          break                          List all breakpoints. (Same as `breakpoints`)
          break --show 2                 Show details about breakpoint #2.
      BANNER

      def options(opt)
        opt.on :c, :condition,     'Change the condition of a breakpoint.', argument: true, as: Integer
        opt.on :s, :show,          'Show breakpoint details and source.',   argument: true, as: Integer
        opt.on :D, :delete,        'Delete a breakpoint.',                  argument: true, as: Integer
        opt.on :d, :disable,       'Disable a breakpoint.',                 argument: true, as: Integer
        opt.on :e, :enable,        'Enable a disabled breakpoint.',         argument: true, as: Integer
        opt.on     :'disable-all', 'Disable all breakpoints.'
        opt.on     :'delete-all',  'Delete all breakpoints.'
        method_options(opt)
      end

      def process
        Pry.processor.pry = _pry_

        {
          delete:        :delete,
          disable:       :disable,
          enable:        :enable,
          'disable-all': :disable_all,
          'delete-all':  :clear,
        }.each do |action, method|
          if opts.present?(action)
            Breakpoints.__send__ method, *(method == action ? [opts[action]] : [])
            return run 'breakpoints'
          end
        end

        if opts.present?(:condition)
          Breakpoints.change(opts[:condition], args.empty? ? nil : args.join(' '))
          run 'breakpoints'
        elsif opts.present?(:show)
          print_full_breakpoint Breakpoints.find_by_id(opts[:show])
        elsif args.empty?
          run 'breakpoints'
        else
          return unless PryDebuggerJRuby.check_trace_enabled
          new_breakpoint
        end
      end

      def new_breakpoint
        place = args.shift
        condition = args.join(' ') if 'if' == args.shift

        file, line =
          case place
          when /^(\d+)$/       # Line number only
            line = $1
            unless PryDebuggerJRuby.check_file_context(target)
              raise ArgumentError, 'Line number declaration valid only in a file context.'
            end
            [target.eval('__FILE__'), line]
          when /^(.+):(\d+)$/  # File and line number
            [$1, $2]
          else               # Method or class name
            self.args = [place]
            method_object.source_location
          end

        print_full_breakpoint Breakpoints.add(file, line.to_i, condition)
      end
    end
    alias_command 'breakpoint', 'break'

    create_command 'breakpoints' do
      description 'List defined breakpoints.'

      banner <<-BANNER
        Usage:   breakpoints [OPTIONS]
        Aliases: breaks

        List registered breakpoints and their current status.
      BANNER

      def options(opt)
        opt.on :v, :verbose, 'Print source around each breakpoint.'
      end

      def process
        if Breakpoints.count > 0
          if opts.verbose?   # Long-form with source output
            Breakpoints.each { |b| print_full_breakpoint(b) }
          else               # Simple table output
            max_width = [Math.log10(Breakpoints.count).ceil, 1].max
            header = "#{' ' * (max_width - 1)}#  Enabled  At "

            output.puts
            output.puts text.bold(header)
            output.puts text.bold('-' * header.size)
            Breakpoints.each do |breakpoint|
              output.printf "%#{max_width}d  ", breakpoint.id
              output.print  breakpoint.enabled? ? 'Yes      ' : 'No       '
              output.print  "#{breakpoint.source}:#{breakpoint.pos}"
              output.print  " (if #{breakpoint.expr})" if breakpoint.expr
              output.puts
            end
            output.puts
          end
        else
          output.puts text.bold('No breakpoints defined.')
        end
      end
    end
    alias_command 'breaks', 'breakpoints'

    helpers do
      def breakout_navigation(action, times = nil)
        return unless PryDebuggerJRuby.check_trace_enabled

        _pry_.binding_stack.clear     # Clear the binding stack
        throw :breakout_nav, {        # Break out of the REPL loop and
          action: action,             #   signal the tracer
          times:  times,
          pry:    _pry_,
        }
      end

      # Ensures that a command is executed in a local file context.
      def check_file_context
        unless PryDebuggerJRuby.check_file_context(target)
          raise Pry::CommandError, 'Cannot find local context. Did you use `binding.pry`?'
        end
      end

      # Print out full information about a breakpoint including surrounding code
      # at that point.
      def print_full_breakpoint(breakpoint)
        line = breakpoint.pos
        output.print text.bold("Breakpoint #{breakpoint.id}: ")
        output.print "#{breakpoint.source} @ line #{line} "
        output.print breakpoint.enabled? ? '(Enabled)' : '(Disabled)'
        output.puts  ' :'
        if (expr = breakpoint.expr)
          output.puts "#{text.bold('Condition:')} #{expr}"
        end
        output.puts
        output.puts(
          Pry::Code.from_file(breakpoint.source)
            .around(line, 3).with_line_numbers.with_marker(line).to_s
        )
        output.puts
      end
    end
  end
end

Pry.commands.import PryDebuggerJRuby::Commands
